Fix the wget footer

Yusuto opened the pull request:
...by having it use the absolute url instead of the breadcrumbs.
I think the changes should be rather self-explanatory, I tested my changes before opening this PR and also ran Clippy and rustfmt and neither said anything, but please still check what I did, just in case.

svenstaro commented on src/renderer.rs:
```diff
@@ -267,11 +267,11 @@ fn version_footer() -> Markup {
-        let count_slashes = title_path.matches('/').count();
-        if count_slashes > 0 {
-            count_slashes - 1
+        let count_slashes = abs_path.matches('/').count();
+        if count_slashes >= 4 {
+            count_slashes - 4
```
That's a bit sketchy. Why 4?

Yusuto replied:
Oh, that change is because `abs_path` has more slashes than the breadcrumbs, since it includes the protocol. But wait, why did I increase it by three, not just two? It should be `http://test.net/my-folder/.git` compared to `test.net/my-folder/.git`, only two more slashes... I made it so the `--cut-dirs` flag has the same value as before the change, but I can't explain exactly why it is that way, Imma investigate

Yusuto replied:
Oh I know why! `abs_path` includes an extra slash at the end which the breadcrumbs didn't, so the two slashes from the protocol plus the one at the end made me increase it by three. But that means that there is now two slashes at the end of the path in the command, I need to fix that.

Yusuto replied:
I think I made the fix and it should be part of the PR now, I made an embarrassing typo in the commit message though which I discovered too late because I wasn't careful before pushing to my fork, I fixed the typo in an amend and force pushed that to my fork, I hope that's not causing any issues.

svenstaro commented:
Please squash your commits.

svenstaro commented:
Fixes #1039.

Yusuto commented:
> Please squash your commits.

I think I squashed my commits, I hope I did what you mean and didn't mess anything up with the force pushes

Yusuto commented on src/renderer.rs:
```diff
@@ -267,11 +267,11 @@ fn version_footer() -> Markup {
-        let count_slashes = title_path.matches('/').count();
-        if count_slashes > 0 {
-            count_slashes - 1
+        let count_slashes = abs_path.matches('/').count();
```
It's for the `--cut-dirs` of the wget command which I think makes it so you don't download into a bunch of unnecessary subfolders, I will add the comment and add it to the last comment (so another force push) in a bit!

Yusuto replied:
Bit of a panicked mess but I think I got it now!

svenstaro commented:
It's not necessary for this MR btw, but if you'd like to, you could write a small test to make sure the behavior for this footer is always functional. :)

Yusuto commented:
> It's not necessary for this MR btw, but if you'd like to, you could write a small test to make sure the behavior for this footer is always functional. :)

Oh I'd love to do that Yip!
Would it make sense to just test the function `wget_footer` by giving it inputs and seeing if it matches a certain output? That's how I'd do it, but I'd consult a testing manual just in case!

Also I couldn't add the tests right now, should I add them in a separate PR, add there here later (squashed or unsquashed?), or for now not at all?

svenstaro commented:
You could also add the tests in this MR. Don't worry, I'll wait. I just want to release something this week ideally. 

I think in this case a unit test of this particular function in the same file makes sense as well as an integration test. we have plenty of examples for both so it shouldn't be that hard. :)

Yusuto commented:
Yip, sounds gud, Thanks! I'll get to it either later today or tomorrow morning :sunny:

Yusuto commented:
I'm done I think!

- [x] Fixed footer
- [x] Made footer click to copy (I didn't pay attention to accessibility though)
- [x] Added some fancy improvements
- [x] Added and updated tests

If any issues arise I'll get to them tomorrow, but if nawt then *I think* this MR is ready to merge!

svenstaro commented on src/listing.rs:
```diff
@@ -163,12 +163,12 @@ pub fn directory_listing(
-        req.connection_info().host(),
-        req.uri()
-    );
+    let abs_uri = http::Uri::builder()
+        .scheme(req.connection_info().scheme())
+        .authority(req.connection_info().host())
+        .path_and_query(req.uri().to_string())
+        .build()
+        .unwrap();
```
Yup, this is much more sane.

svenstaro commented on src/renderer.rs:
```diff
@@ -683,3 +695,54 @@ pub fn render_error(
+    use super::*;
+    use pretty_assertions::assert_eq;
+    #[test]
+    fn test_wget_footer() {
+        let to_html = |x| {
+            format!(
+                r#"<div class="downloadDirectory"><p>Download folder:</p><a class="cmd" title="Click to copy!" style="cursor: pointer;" onclick="navigator.clipboard.writeText(&quot;wget -rcnHp -R 'index.html*' {0}/?raw=true'&quot;)">wget -rcnHp -R 'index.html*' {0}/?raw=true'</a></div>"#,
+                x
+            )
+        };
+        let uri = |x| Uri::try_from(x).unwrap();
+
+        let to_be_tested: String = wget_footer(
+            &uri("https://github.com/svenstaro/miniserve/"),
+            Some("Miniserve"),
+            None,
```
Why not make all of these test cases into their own unit tests? That way, you can also make it self-documenting by varying the test names.

svenstaro commented on src/renderer.rs:
```diff
@@ -267,26 +268,37 @@ fn version_footer() -> Markup {
-            0
-        }
+fn wget_footer(abs_path: &Uri, root_dir_name: Option<&str>, current_user: Option<&str>) -> Markup {
+    let escape_apostrophes = |x: &str| x.replace('\'', "'\"'\"'");
```
Let's not do the JS fat arrow function thing here. Why not just make this a proper little inner function instead?

svenstaro commented on src/renderer.rs:
```diff
@@ -267,26 +268,37 @@ fn version_footer() -> Markup {
+        // Put all the files in a folder of this name
+        0 => format!(
+            " -P '{}'",
+            escape_apostrophes(
```
How do you run into the case that this is required anyway?

svenstaro commented:
Hey, do you plan on fixing up the rest soonish? I'd like to do a release. No worries if you need longer but then it might not make the release.

Yusuto commented:
> Hey, do you plan on fixing up the rest soonish? I'd like to do a release. No worries if you need longer but then it might not make the release.

Oh, I'm not exactly sure what is left to be fixed, it's only the tests, right?

I wrote a comment earlier that I'm very unsure about how to split them apart, whatever approach I try seems to be just kind of not really optimal, or at least not significantly better than just lumping it all in one function and if something fails manually figuring out what it is, so I'm a bit at loss of how to continue there I'm afraid

svenstaro commented:
Merging as is, thanks! I split the tests the way I imagined it.

Yusuto commented:
Oké, Thank you! This is my first MR I ever made, I'm very happy to get it accepted ^^

svenstaro commented:
Thanks for the contribution :) I'd be happy for more contributions of course.

Yusuto commented:
No problem!
I will definitely consider it, Yeh! More advanced uploading (like uploading folders and seeing progress) would be pretty heccin practical I think for example! Although I'm also very interested in media processing and am currently looking into that, I'll also look at issues if I get the time ^^
